### PR TITLE
Feature/uncertainty map

### PR DESCRIPTION
**Pull request by renezurbruegg**
- Created Uncertainty Map 
- Created Voxel Value Evaluator 
- Added parameter to only sample goal point with RRT in the same xy plane

### CODE REVIEW

File: active_3d_planning_core/src/module/trajectory_generator/rrt.cpp [hunk 1]
```diff
@@ -206,7 +210,11 @@ bool RRT::sampleGoal(Eigen::Vector3d* goal_pos) {
     if (semilocal_count_ > 0) {
       // Not enough local points found, sample from local sphere
       double theta = 2.0 * M_PI * static_cast<double>(rand()) / RAND_MAX;
-      double phi = acos(1.0 - 2.0 * static_cast<double>(rand()) / RAND_MAX);
+      double phi = M_PI / 2;
```
Comment (Schmluk) [inline active_3d_planning_core/src/module/trajectory_generator/rrt.cpp:213]: Unnecessary int to double conversion (better use M_PI/2.0)

File: active_3d_planning_voxblox/include/active_3d_planning_voxblox/map/uncertainty_aware_map.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  // implement virtual methods
+  void setupFromParamMap(Module::ParamMap* param_map) override;
+
+  // check collision for a single pose
+  bool isTraversable(const Eigen::Vector3d& position,
+                     const Eigen::Quaterniond& orientation) override;
+
+  // check whether point is part of the map
+  bool isObserved(const Eigen::Vector3d& point) override;
```
Comment (Schmluk) [inline active_3d_planning_voxblox/include/active_3d_planning_voxblox/map/uncertainty_aware_map.h:30]: The vast majority of these functions will be identical to the parent implementation, then you don't need to override them.

File: active_3d_planning_voxblox/src/map/uncertainty_aware_map.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+voxblox::EsdfServer& UncertaintyAwareMap::getESDFServer() {
+  return voxbloxMap.getESDFServer();
+}
+
+void UncertaintyAwareMap::setupFromParamMap(Module::ParamMap* param_map) {
+  // Reconstruction + collision map
+  voxbloxMap.setupFromParamMap(param_map);
+
+  // Uncertainty Map
+  ros::NodeHandle nh("/uncertainty/map/");
```
Comment (Schmluk) [inline active_3d_planning_voxblox/src/map/uncertainty_aware_map.cpp:25]: Avoid global namespaces, this namespace could e.g. be a param of the map.

**Review by Schmluk — APPROVED** [review]
LGTM, see few comments below.

Overall, the new things you develop don't need to be part of the active_3d_planning repo, e.g. you can also have your app_uncertainty package in your private repo and extend active_3d in the same way you already did. Otherwise it will be public and might also be confusing for people (in particular if there's some experimental / WIP code). Things that are generally useful can be added to active_3d_planning but should then also be listed in the wiki etc when merged to master.

File: active_3d_planning_core/include/active_3d_planning_core/module/trajectory_evaluator/voxel_value_evaluator.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+#include "active_3d_planning_core/module/trajectory_evaluator/frontier_evaluator.h"
+
+namespace active_3d_planning {
+namespace trajectory_evaluator {
+
+// VoxelValue uses a value stored in the tsdf of surface voxels as gain to
+// evaluate a given trajectory.
+// Requires the map to be an instance of VoxelValueMap.
+// Uses frontier voxels to allow additional gain for exploration.
+class VoxelValueEvaluator : public FrontierEvaluator {
```
[the remark this answers is not in the document]
Comment (renezurbruegg) [reply active_3d_planning_core/include/active_3d_planning_core/module/trajectory_evaluator/voxel_value_evaluator.h:17]: I thought i would be nice to have the additional functionality of having a different weight for frontier voxels

File: active_3d_planning_voxblox/src/map/uncertainty_aware_map.cpp (hunk 3 above)
Comment (renezurbruegg) [reply active_3d_planning_voxblox/src/map/uncertainty_aware_map.cpp]: cc32191

File: active_3d_planning_core/include/active_3d_planning_core/module/trajectory_generator/rrt.h [hunk 5]
```diff
@@ -82,7 +82,7 @@ class RRT : public TrajectoryGenerator {
   bool p_sample_yaw_;    // True: random yaw, false: face direction of travel
   int p_maximum_tries_;  // sampling tries, 0 for inf
   double p_min_path_length_;  // disallow all segments shorter than min length
-
+  bool planar; // Only sample in xy plane
```
[the remark this answers is not in the document]
Comment (renezurbruegg) [reply active_3d_planning_core/include/active_3d_planning_core/module/trajectory_generator/rrt.h]: c800ae7

**Pull request merged**